### DISCUSSION TITLE
Combine matrices

### DISCUSSION THREAD

**Discussion by raphaelmenges** [dimforge/nalgebra#1420] [Q&A] [answered]
Hello 👋,

I looked through the documentation and search but could not find a good example how to combine multiple matrices or assign or replace a whole block of an existing matrix in `nalgebra`. E.g., I have a rotation matrix (3x3) and a translation vector (3x1) and I want to put them together as transformation matrix (3x4). What is the advisable way to achieve this?

**Comment by Andlon** [comment, marked as answer]
Hey @raphaelmenges. `nalgebra` 0.33 was released a few days ago, and it includes the new [`stack!`](https://docs.rs/nalgebra/latest/nalgebra/macro.stack.html) macro, which hopefully offers you a convenient way to solve your problem.

**Reply by raphaelmenges** [reply]
Great, thanks for the reply! I am now actually trying to compose a 4x4 transformation matrix with homogeneous coordinates. I found two ways:

```rust
stack![rotation_and_scale, translation;
        matrix![0.,0.,0.], matrix![1.]]
```

or

```rust
stack![stack![rotation_and_scale, translation];
        matrix![0.,0.,0.,1.]]
```

Sadly, I cannot:

```rust
stack![rotation_and_scale, translation;
        matrix![0.,0.,0.,1.]]
```

Which way to prefer? I guess the second approach makes additional copies in-between 🤔

**Reply by Andlon** [reply]
I'd probably do something like this, personally:

```rust
let one = matrix![ 1.0 ];
let matrix = stack![ rotation_and_scale, translation;
                                      0,         one ];
```
This uses a nifty feature of `stack!`, in which `0` literals are interpreted as a zero matrix of the appropriate size (see the docs for more info).
